pkg/alpine/Dockerfile

```diff
@@ -1,4 +1,5 @@
-FROM lfedge/eve-alpine-base:f11c7a5dcffe1ab7b466bea7e26010a27da62b3b AS cache
+# image was bootstraped using FROM lfedge/eve-alpine-base:f11c7a5dcffe1ab7b466bea7e26010a27da62b3b AS cache
+FROM lfedge/eve-alpine:145f062a40639b6c65efa36bed1c5614b873be52 AS cache
```

Comment:
I do not understand this. Isn't this file (`pkg/alpine/Dockerfile`) the source of `lfedge/eve-alpine`? How can we make an image be `FROM` itself?

In theory, we could make it be `FROM` an earlier version of itself, but that makes it quite hard to rebuild and reproduce.

What do we gain by moving it to this, as opposed to `FROM lfedge/eve-alpine-base`?

Reply:
`eve-alpine-base` contains only several packages required to run alpine (it based on minirootfs of alpine). We download packages from repository and store them cache (inside `/mirror` directory) during building of `eve-alpine`. We [check](https://github.com/lf-edge/eve/blob/5729285b235dfa09b8704ce758e4fbb4882b76fa/pkg/alpine/build-cache.sh#L19-L22) that package are inside the cache and download them only if they are not cached. So, we should use previous build here as a base to not download new version again, that is, pin them.

Reply:
This sounds like an infinite loop. I am sure it isn't but it sounds like it.

```
eve-alpine-base ---> eve-alpine ---> eve-alpine ---> eve-alpine ---> ... upstream packages
```

How do you break out of it? How do you recreate it? Do I have to check out an earlier version of this repo, build one thing, then check out a later and rebuild? That isn't an easily understood design.

Why would we not just have:

```
eve-alpine-base ---> eve-alpine ---> upstream packages
```

Reply:
I still don't get it; I must be missing something.

Isn't the whole purpose of eve-alpine to have a cache, i.e. that is where we pin everything? eve-alpine _is_ the cache for upstream packages. Why do I need another cache layer beyond that?

I get that you need `eve-alpine-base`, even if you would prefer not to, because of the lack of riscv64 support. If `library/alpine` had full riscv64 support, you wouldn't need `eve-alpine`.

But isn't the purpose of `eve-alpine` to download everything we need upstream into a cache in the image? And then we use that as our "package source" in `upstream`?

Reply:
Imagine that you want to add new package in `eve-alpine`. In that case you will add the package to the list [here](https://github.com/lf-edge/eve/tree/master/pkg/alpine/mirrors). And you will need to build new version of `eve-alpine` with desired package inside. If we will use `FROM lfedge/eve-alpine-base` or `FROM library/alpine` we will download all packages again and your new version of `eve-alpine` will contains new `/mirror` with potentially new versions of packages. You will use your new image only in one package you modify and potentially new package versions will not hurt your workflow. But next time you will want to sync `FROM lfedge/eve-alpine` line in other packages you may hit an error for example with gcc version in kernel, which is not expected.

Reply:
This (good) comment of yours is the first I have seen of a problem definition. I am going to try and restate it, so go ahead and correct it. Once it is correct, it should be the beginning of the description of the PR (or an issue that this PR links to).

1. We had a problem that every time we rebuild any downstream `pkg/` images, alpine apk packages could change.
2. We solved the above by caching everything in `eve-alpine` as a base, which everything takes from.

Since downstream change very frequently, and eve-alpine infrequently, as long as `eve-alpine` doesn't change, or the downstream doesn't use a different `FROM` hash, then downstream is guaranteed the same bits for a given `apk` package every time.

First problem solved.

However, we have a second-order problem (one I think you are outlining here).

1. downstream image needs to add a new package
2. to do so, we update the base image with its cache, rebuild it, and make downstream depend on new version
3. because we rebuild cache, bits of other packages (not just the newly added on) can change

You are trying to find a way (with this PR) to be able to modify or add just the one package?

Reply:
Correct. We solve the first problem just having eve-alpine. And we solve the second problem using previous eve-alpine as a base for the current (i.e. what we [have](https://github.com/lf-edge/eve/blob/8.10.0/pkg/alpine/Dockerfile#L1-L2) in the last release 8.10.0).

Reply:
A few things.

First, the description of the problem that you are trying to solve definitely should be **the** primary description of this PR. The most important thing is for people to understand the problem you are solving, and how.

Second, I am not sure I understand how this solves it. Let's play it out.

I have ten packages, call them A, B, C, D, E, F, G, H, I, J. Each is at version 1.0.0, and is now in the cache.

I decide to add package Q (non-sequential makes life more interesting, and Q does the gadgets for James Bond). I add it to `mirrors/3.16/main` and rerun the build. It then runs `apk add` for each package in `main`. Won't that cause them all to reinstall anyways?

Or are you saying that because `apk add` always looks in its cache, and because it sees all of the others `A`-`J` in there, it only will add `Q`? If that is correct, the design should be described here in the PR, but (even more importantly) in the docs (either in the `pkg/alpine/` directory or in `docs/`), so people coming to it in the future understand the design, how it works, and why it works this way.

Also, if that is correct, what happens if I want to get a newer version of package `C`? How do I do that?

If I understood it correctly, this becomes a chained cache: each version of the cache builds on the previous version. Will we hit any limits in docker because of it? What will it look like? How will I reproduce the image from scratch?

Either way, there is a lot of (quite impressive) fundamental design going on here; I still don't understand some parts, but there really big thing missing is some solid documentation loaded into the repo: how does this work? what are the scenarios and how do I run them (add a package, update an existing one, update alpine version, etc.)? what assumptions are we making? how do I rebuild the whole thing from scratch? how do I avoid layer limits (if there are any in this scenario)?

Either way, this is a lot of convoluted engineering to get around the fact that I simply cannot `apk add <something>` and have it reliably do the same bits over and over, as I can do `docker pull foo@sha256:<hash>`. If we could do `apk add foo=1.2.3@sha256:<hash>`, all of this would disappear, correct?

Reply:
> A few things.
> 
> First, the description of the problem that you are trying to solve definitely should be **the** primary description of this PR. The most important thing is for people to understand the problem you are solving, and how.

Strictly speaking I just want to rollback to the logic we have for years after https://github.com/lf-edge/eve/pull/519 merged. We have the documentation here: https://github.com/lf-edge/eve/blob/master/docs/BUILD.md#how-to-update-eve-alpine-package

> Second, I am not sure I understand how this solves it. Let's play it out.
> 
> I have ten packages, call them A, B, C, D, E, F, G, H, I, J. Each is at version 1.0.0, and is now in the cache.
> 
> I decide to add package Q (non-sequential makes life more interesting, and Q does the gadgets for James Bond). I add it to `mirrors/3.16/main` and rerun the build. It then runs `apk add` for each package in `main`. Won't that cause them all to reinstall anyways?

As I mentioned in previous comments: we [check](https://github.com/lf-edge/eve/blob/5729285b235dfa09b8704ce758e4fbb4882b76fa/pkg/alpine/build-cache.sh#L19-L22) that package is inside the cache and download them only if they are not cached. But let me check what happens with dependencies of the package we try to install, it is no clear for me, as we first try to fetch with `recursive` flag [here](https://github.com/lf-edge/eve/blob/5729285b235dfa09b8704ce758e4fbb4882b76fa/pkg/alpine/build-cache.sh#L27).

> Also, if that is correct, what happens if I want to get a newer version of package `C`? How do I do that?

Good question. And also we cannot remove the package (for example to reduce the size). I think in that case we should somehow remove the apk file from cache before we run `build-cache.sh`. As a first implementation we can add empty list of packages we want to remove from cache and comment about the process into Dockerfile of eve-alpine. But It really looks like separate issue and separate PR.

> If I understood it correctly, this becomes a chained cache: each version of the cache builds on the previous version. Will we hit any limits in docker because of it? What will it look like?

You are right. I hope I resolved the problem https://github.com/lf-edge/eve/pull/2721. But probably we should completely remove and replace old `/mirror` if we want to support delete/update operations (again, not sure that we want to implement them now).

> How will I reproduce the image from scratch?

You will need to download something in all cases. If FROM lfedge/eve-alpine will not fit your requirements, you can see my comment in the first line of Dockerfile and use FROM lfedge/eve-alpine-base instead.

> Either way, there is a lot of (quite impressive) fundamental design going on here

Not sure, we've had this design for almost three year now.

> Either way, this is a lot of convoluted engineering to get around the fact that I simply cannot `apk add <something>` and have it reliably do the same bits over and over, as I can do `docker pull foo@sha256:<hash>`. If we could do `apk add foo=1.2.3@sha256:<hash>`, all of this would disappear, correct?

I think even `apk add foo=1.2.3` will meet our requirements, but alpine's repo contains only the last versions, and tomorrow our build will fail.

Reply:
Yes, adding a package can be problematic if that foo=1.2.3 does not exist or assume/require new versions of other packages. That will require some more manual care to inspect which packages were updated in lfedge/eve-alpine-base as a result of adding foo. And if version 1.2.3 doesn't exist in Alpine 3.16 then we'd need to move alpine forward to a new version. Nothing surprising here AFAICT.

So far we haven't had a need to remove any packages from eve-alpine (but if we get to FIPS-140 we presumably want to make sure eve-alpine and eve-alpine-base do not contain any crypto packages we do not need and use to make the compliance/certification easier, thus we might need to delete any superflous *tls* and *ssl* packages). But we can cross that bridge when we get to it.

Reply:
> Strictly speaking I just want to rollback to the logic we have for years after https://github.com/lf-edge/eve/pull/519 merged

That whole thing was quite confusing; I never understood how the Dockerfile to `lfedge/eve-alpine` could be `FROM lfedge/eve-alpine`! I think the comments here may be the first location it is clearly described.

> Good question. And also we cannot remove the package (for example to reduce the size). I think in that case we should somehow remove the apk file from cache before we run build-cache.sh. As a first implementation we can add empty list of packages we want to remove from cache and comment about the process into Dockerfile of eve-alpine. But It really looks like separate issue and separate PR.

Our design (which was in place, as you pointed out, but very confusing, then removed, possibly by chance, now you want to put back) apparently had some functional weaknesses, not just design questions.

Let me get this straight: unlike container images (which, admittedly, is fairly new, but younger than Alpine IIRC), where I always can get a certain guaranteed version by using hashes, and tags are not really supposed to be mutable but are, with Alpine I have no way of saying, "give me this precise and unchanged version of package X"; "unchanged" = "exact same bits if I ask for the same version of the package over and over again".

That frustrating design is at the heart of our problems here.

So we built a cache. Then we built a way to add on only individual packages of the cache, because rebuilding the whole cache is inefficient and might change for the same reason (not very "cloud-native"). Then we are realizing that updating individual packages in the cache is nearly impossible and may need even more changes. Then we are realizing that removing packages from the cache may require a whole other layer of engineering.

All of this, when all you really want is to be able to say: "here is a list of packages and their versions. Every time I run build, please ensure that they are the same, except for those I have changed (updated version numbers) or added, and if something is not there, please remove it."

Isn't that what you really are after?